engine/src/witness/sol/program_swaps_witnessing.rs

```diff
@@ -0,0 +1,508 @@
+
+pub async fn get_program_swaps(
+	sol_rpc: &SolRetryRpcClient,
+	swap_endpoint_data_account_address: SolAddress,
+	sc_open_accounts: Vec<SolAddress>,
+	sc_closure_initiated_accounts: BTreeSet<VaultSwapAccountAndSender>,
+	usdc_token_mint_pubkey: SolAddress,
+) -> Result<
+	(Vec<(VaultSwapAccountAndSender, SolanaVaultSwapDetails)>, Vec<VaultSwapAccountAndSender>),
+	anyhow::Error,
+> {
+	let (new_program_swap_accounts, closed_accounts, slot) = get_changed_program_swap_accounts(
+		sol_rpc,
+		sc_open_accounts,
+		sc_closure_initiated_accounts,
+		swap_endpoint_data_account_address,
+	)
+	.await?;
+
+	if new_program_swap_accounts.is_empty() {
+		return Ok((vec![], closed_accounts));
+	}
+
+	let new_swaps = stream::iter(new_program_swap_accounts)
+		.chunks(MAX_MULTIPLE_EVENT_ACCOUNTS_QUERY)
+		.map(|new_program_swap_accounts_chunk| {
+			get_program_swap_event_accounts_data(sol_rpc, new_program_swap_accounts_chunk, slot)
+		})
+		.buffered(MAXIMUM_CONCURRENT_RPCS)
+		.map_ok(|program_swap_account_data_chunk| {
+			stream::iter(program_swap_account_data_chunk.into_iter().filter_map(
+				|(account, program_swap_account_data)| match program_swap_account_data {
+					Some(data)
+						if (data.src_token.is_none() ||
+							data.src_token.is_some_and(|addr| addr == usdc_token_mint_pubkey.into())) => {
+
+								let (deposit_metadata, vault_swap_parameters) = match data.ccm_parameters {
+									None => {
+										let CfParameters { ccm_additional_data: (), vault_swap_parameters } =
+											CfParameters::decode(&mut &data.cf_parameters[..]).map_err(|e| warn!("error while decoding CfParameters for solana vault swap: {}. Omitting swap", e)).ok()?;
+										(None, vault_swap_parameters)
+									},
+									Some(ccm_parameters) => {
+										let CfParameters { ccm_additional_data, vault_swap_parameters } =
+											CcmCfParameters::decode(&mut &data.cf_parameters[..]).map_err(|e| warn!("error while decoding CcmCfParameters for solana vault swap: {}. Omitting swap", e)).ok()?;
+
+										let deposit_metadata = Some(CcmDepositMetadata {
+											source_chain: cf_primitives::ForeignChain::Solana, // TODO: Pass chain id from above?
+											source_address: Some(ForeignChainAddress::Sol(data.sender.into())),
+											channel_metadata: CcmChannelMetadata {
+												message: ccm_parameters.message
+													.to_vec()
+													.try_into()
+													.map_err(|_| anyhow!("Failed to deposit CCM: `message` too long.")).ok()?,
+												gas_budget: ccm_parameters.gas_amount.into(),
+												ccm_additional_data,
+											},
+										});
+										(deposit_metadata, vault_swap_parameters)
+									}
+								};
+
+								Some(Ok((VaultSwapAccountAndSender {
+									vault_swap_account: account,
+									swap_sender: data.sender.into()
+								}, SolanaVaultSwapDetails {
+									from: if data.src_token.is_none() {SolAsset::Sol} else {SolAsset::SolUsdc},
```

Comment:
this is pretty high risk - I would say we explicitly check that it's USDC here - I know we check it above, but seems something easy to miss - or ideally structure the code so it's clear. like have a `get_src_asset -> Option<asset>` function on the data... don't think this works super well, since the `usdc_token_mint_pubkey` is only available in this function's context, but yeah, worth seeing if we can make it a bit clearer/more explicit

Reply:
changed so that we do the check at one place and use the Asset from there.